ndb/src/google/cloud/ndb/model.py

```diff
@@ -2241,9 +2252,240 @@ def __init__(self, *args, **kwargs):
 
-    def __init__(self, *args, **kwargs):
+    The constructor for :class:`KeyProperty` allows at most two positional
+    arguments. Any usage of :data:`None` as a positional argument will
+    be ignored. Any of the following signatures are allowed:
+
+    .. testsetup:: key-property-constructor
+
+        from google.cloud import ndb
+
+
+        class SimpleModel(ndb.Model):
+            pass
+
+    .. doctest:: key-property-constructor
+
+        >>> name = "my_value"
+        >>> ndb.KeyProperty(name)
+        KeyProperty(b'my_value')
+        >>> ndb.KeyProperty(SimpleModel)
+        KeyProperty(kind=b'SimpleModel')
+        >>> ndb.KeyProperty(name, SimpleModel)
+        KeyProperty(b'my_value', kind=b'SimpleModel')
+        >>> ndb.KeyProperty(SimpleModel, name)
+        KeyProperty(b'my_value', kind=b'SimpleModel')
+
+    The type of the positional arguments will be used to determine their
+    purpose: a string argument is assumed to be the ``name`` and a
+    :class:`type` argument is assumed to be the ``kind`` (and checked that
+    the type is a subclass of :class:`Model`).
+
+    .. automethod:: _validate
+
+    Args:
+        name (str): The name of the property.
+        kind (Union[type, str]): The (optional) kind to be stored. If provided
+            as a positional argument, this must be a subclass of :class:`Model`
+            otherwise the kind name is sufficient.
+        indexed (bool): Indicates if the value should be indexed.
+        repeated (bool): Indicates if this property is repeated, i.e. contains
+            multiple values.
+        required (bool): Indicates if this property is required on the given
+            model type.
+        default (.Key): The default value for this property.
+        choices (Iterable[.Key]): A container of allowed values for this
+            property.
+        validator (Callable[[~google.cloud.ndb.model.Property, .Key], bool]): A
+            validator to be used to check values.
+        verbose_name (str): A longer, user-friendly name for this property.
+        write_empty_list (bool): Indicates if an empty list should be written
+            to the datastore.
+    """
+
+    _kind = None
+
+    def __init__(
+        self,
+        *args,
+        name=None,
+        kind=None,
+        indexed=None,
+        repeated=None,
+        required=None,
+        default=None,
+        choices=None,
+        validator=None,
+        verbose_name=None,
+        write_empty_list=None
+    ):
+        name, kind = self._handle_positional(args, name, kind)
+        super(KeyProperty, self).__init__(
+            name=name,
+            indexed=indexed,
+            repeated=repeated,
+            required=required,
+            default=default,
+            choices=choices,
+            validator=validator,
+            verbose_name=verbose_name,
+            write_empty_list=write_empty_list,
+        )
+        if kind is not None:
+            self._kind = kind
```

Comment:
As in all subtypes that add a data attribute, I'm making this optional even though the original implementation would always set `self._kind` even if `kind` were `None`

Reply:
Is this useful without a kind? Should it be required?

Reply:
@chrisrossi Did you resolve your

> Is this useful without a kind? Should it be required?

question? (I saw it in email but not on GitHub.)

Reply:
I decided after looking at it more that it probably was useful and retracted my question.